### PR TITLE
Transition future compat lints to {ERROR, DENY} - Take 2

### LINKED ISSUE #36247: Tracking issue for `safe_extern_statics` compatibility lint
#### What is this lint about

Consider the following code:

```
extern {
    static S: T;
}
```

`S` here is a static variable accessible to Rust but defined and initialized elsewhere.
External static variables cannot be accessed safely from Rust code for several reasons:
- `S` can contain an arbitrary bit pattern not necessarily being a valid value of type `T`.
- even if all bit patterns are valid values of `T`, `S` is still not controlled by Rust ownership system and can be accessed from other threads potentially creating data races.

See #35112 for more details as well as examples of how this can lead to crashes.

Previously safe accesses to extern statics were erroneously permitted outside of unsafe blocks. https://github.com/rust-lang/rust/pull/36173 fixed this oversight.
#### How to fix this warning/error

Surround accesses to extern statics with `unsafe` blocks and make sure these accesses are actually valid.
#### Current status
- [x] https://github.com/rust-lang/rust/pull/36173 introduces the `safe_extern_statics` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/42894 makes the `safe_extern_statics` lint deny-by-default
- [ ] PR ? makes the `safe_extern_statics` lint a hard error

### LINKED ISSUE #37872: `legacy_directory_ownership` future-compatibility warnings
This is the summary issue for the `legacy_directory_ownership` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made.

#### What is the warning for?

The `legacy_directory_ownership` warning is issued when
 - there is a non-inline module with a `#[path]` attribute (e.g. `#[path = "foo.rs"] mod bar;`),
 - the module's file ("foo.rs" in the above example) is not named "mod.rs", and
 - the module's file contains a non-inline child module without a `#[path]` attribute.

The warning can be fixed by renaming the parent module to "mod.rs" and moving it into its own directory if appropriate.

We are making this a warning so that the directory ownership rules apply more consistently to modules with `#[path]` attributes. This is described in more detail in #37602.

#### When will this warning become a hard error?

At the beginning of each 6-week release cycle, the Rust compiler team will review the set of outstanding future compatibility warnings and nominate some of them for Final Comment Period. Toward the end of the cycle, we will review any comments and make a final determination whether to convert the warning into a hard error or remove it entirely.

#### Current status
- [x] https://github.com/rust-lang/rust/pull/37602 introduces the `legacy_directory_ownership` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/42894 makes the `legacy_directory_ownership` lint deny-by-default
- [ ] PR ? makes the `legacy_directory_ownership` lint a hard error

### LINKED ISSUE #39207: Tracking issue for `legacy_constructor_visibility` compatibility lint
#### What is this lint about

https://github.com/rust-lang/rust/pull/38932 introduces new rules for privacy of tuple struct constructors. Constructor is a function automatically created by a tuple struct having the same name as the struct itself.
```
struct S(u8); // defines `struct S { ... }` and `fn S(u8) -> S` 
```
See [RFC 1506](https://github.com/rust-lang/rfcs/blob/master/text/1506-adt-kinds.md) for more details.

With these new rules constructor function of a public `struct S(....)` is defined as private if at least one of the fields is private.
More precisely, `visibility(constructor_fn) = min(visibility(struct), visibility(field_1), ..., visibility(field_N))` (this definition takes [RFC 1422](https://github.com/rust-lang/rfcs/blob/master/text/1422-pub-restricted.md) into account).

This change is done to make a struct with private fields being a tuple struct an implementation detail. So libraries can freely change it into non-tuple struct and back, and nothing breaks.
This change is *mostly* backward compatible due to some ad-hoc privacy checks existing previously (e.g. `error[E0450]: cannot invoke tuple struct constructor with private fields`), however there's one pattern that is broken by it:

```
mod m {
    // Tuple struct with a private field.
    // Type S is pub(pub).
    // The field S::0 is pub(m).
    // Constructor S is min(pub(pub), pub(m)) -> pub(m).
    pub struct S(u8);
    
    fn f() {
        // Try to use S from the root module in value namespace.
        // No success, ::S exists only in type namespace.
        // How to fix: use S from this module instead of ::S from the root module.
        ::S;
    }
}

// This imports S only in type namespace, value S is too private.
// This is expected filtering behavior of imports described in RFC 1560.
use m::S;

fn main() {}
```

`legacy_constructor_visibility` lint tries to detect this patterns and make name resolution succeed with a warning to keep backward compatibility. However, this detection is pretty hacky and not entirely precise, so it will need to go away eventually.

#### How to fix this warning/error

Use the constructor from its original location X and not through reexports in modules outer to X.

#### Current status
- [x] https://github.com/rust-lang/rust/pull/38932 introduces the `legacy_constructor_visibility` lint as deny-by-default
- [ ] PR ? makes the `legacy_constructor_visibility` lint a hard error

### LINKED ISSUE #42238: `parenthesized_params_in_types_and_modules` future-compatibility warnings
This is the **summary issue** for the `PARENTHESIZED_PARAMS_IN_TYPES_AND_MODULES` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made.

#### What is the warning for?

The `parenthesized_params_in_types_and_modules` warning is issued when you use parenthesized generic parameters in modules, primitive types, type parameters, or `Self`. For structs, enums, enum variants, unions, and type aliases, the syntax has already issued an error. Typical example of this warning is a code like this:

```rust
fn something(&self) -> i32 () {
  //                       ^^ parenthesized params

  ..
  let x = self.foo() as usize();
  //                         ^^ parenthesized params
  ..
}
```

This warning can always be fixed by removing the parentheses.

#### When will this warning become a hard error?
                    
At the beginning of each 6-week release cycle, the Rust compiler team will review the set of outstanding future compatibility warnings and nominate some of them for **Final Comment Period**. Toward the end of
the cycle, we will review any comments and make a final determination whether to convert the warning into a hard error or remove it entirely.

#### Current status
- [x] https://github.com/rust-lang/rust/pull/41856 introduces the `parenthesized_params_in_types_and_modules` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/42894 makes the `parenthesized_params_in_types_and_modules` lint deny-by-default
- [ ] PR ? makes the `parenthesized_params_in_types_and_modules` lint a hard error

### LINKED ISSUE #59014: Tracking issue for `nested_impl_trait` compatibility lint
#### What is this lint about

Some occurrences of `impl Trait` that appear nested at certain points within the type structure of a larger `impl Trait` type are meant to be rejected by the compiler, because (at least in part) the assignment of concrete types to those nested `impl Trait` is [fragile and fails to provide a path for the user to be more explicit should inference fail][fragile].

[fragile]: https://github.com/rust-lang/rust/issues/57979#issuecomment-459387604

Due to implementation oversights, some occurrences were erroneously skipped by the nested `impl Trait` detection, and then those occurrences were sometimes successfully assigned to a type based on Rust type inference implementation details. These implementation oversights were long overlooked and thus became part of the stable Rust compiler.

PR #57730 fixed those aforementioned implementation details, and thus some instances of `impl Trait` are rejected as erroneous when they were previously accepted by the stable compiler. Since our policy is to provide a smooth transition when such errors are introduced (see [RFC 1589][]), PR #58608 adapted those fixes to the implementation to downgrade those particular rejections of nested `impl Trait` to be linted as warnings rather than errors.

[RFC 1589]: https://github.com/rust-lang/rfcs/blob/master/text/1589-rustc-bug-fix-procedure.md

#### How to fix this warning/error

All cases where this code was known to successfully compile are instances of `impl Trait` in argument position (#44721). All such cases can be rewritten to instead use explicit generic type parameters. For example, this signature from #57979 

```rust
pub fn collect(_: impl IntoIterator<Item = impl Borrow<Data<impl AsRef<[u8]>>>>)
```

can be rewritten as follows:

```rust
pub fn collect<T>(_: impl IntoIterator<Item = impl Borrow<Data<T>>>) where T: AsRef<[u8]>
```

#### Current status
- [x] https://github.com/rust-lang/rust/pull/58608 introduces the `nested_impl_trait` lint as warn-by-default
- [ ] PR ? makes the `nested_impl_trait` lint deny-by-default
- [ ] PR ? makes the `nested_impl_trait` lint a hard error

### PR DESCRIPTION
**Pull request by Centril**
Follow up to https://github.com/rust-lang/rust/pull/63247 implementing https://github.com/rust-lang/rust/pull/63247#issuecomment-536295992.

- `legacy_ctor_visibility` (ERROR) -- closes #39207
- `legacy_directory_ownership` (ERROR) -- closes #37872
- `safe_extern_static` (ERROR) -- closes #36247
- `parenthesized_params_in_types_and_modules` (ERROR) -- closes #42238
- `duplicate_macro_exports` (ERROR)
- `nested_impl_trait` (ERROR) -- closes #59014
- `ill_formed_attribute_input` (DENY) -- transitions #57571
- `patterns_in_fns_without_body` (DENY) -- transitions #35203

r? @varkor 
cc @petrochenkov

### CODE REVIEW

**Comment by Centril** [comment]
The PR builder is finally happy. 🎉

File: src/libsyntax/attr/builtin.rs [hunk 1]
```diff
@@ -945,11 +945,6 @@ pub fn check_builtin_attribute(
     // Some special attributes like `cfg` must be checked
     // before the generic check, so we skip them here.
     let should_skip = |name| name == sym::cfg;
-    // Some of previously accepted forms were used in practice,
-    // report them as warnings for now.
-    let should_warn = |name| name == sym::doc || name == sym::ignore ||
-                             name == sym::inline || name == sym::link ||
-                             name == sym::test || name == sym::bench;
```
Comment (petrochenkov) [inline src/libsyntax/attr/builtin.rs]: Too soon, IMO.
This is going directly from a warning to hard error and is going to cause a lot of [indirect fallout](https://github.com/rust-lang/rust/pull/57321#issuecomment-451685517).
I'd rather keep the whitelist (while still making the lint deny-by-default).

**Comment by petrochenkov** [comment]
Also, `legacy_ctor_visibility` had a large number of regressions in the crater run in https://github.com/rust-lang/rust/pull/63247, but most of them from unstable `syntex_syntax` (expected function, found struct `ast::Name`), so it should be ok.

File: src/libsyntax/attr/builtin.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/attr/builtin.rs]: Done.

**Comment by Centril** [comment]
@petrochenkov @varkor This should be ready again for review now.

**Comment by Centril** [comment]
Rebased.

**Comment by JohnCSimon** [comment]
Ping from triage.
@varkor  Can you please review this PR?
cc: @Centril 

Thank you!

**Comment by Centril** [comment]
r? @oli-obk 

🤞

**Comment by oli-obk** [comment]
should we have a team sign off on this?

**Comment by Centril** [comment]
I'm a bit biased, but imo the process overhead is unnecessary as this is merely implementing [a previously suggested plan](https://github.com/rust-lang/rust/pull/63247#issuecomment-536295992) and we've done [similar transitions](https://github.com/rust-lang/rust/pull/59928#issuecomment-488664195) before without said overhead.

**Comment by oli-obk** [comment]
I'm all for this, but the precedent also had fix-PRs open for the crates that will break and did a reverse dependency analysis.

I know that doing so is a lot of work, but we do have our stability story. Picking an arbitrary threshold (like 5) of broken crates where it's ok to go to an error, seems a bit odd to me. Then again, we have to do this PR at some point and the forward compat lints have been triggering for a while.

I'm unsure how to proceed, thus my question about talking this through in a team meeting or just giving everyone a chance to sign off or comment.

**Comment by oli-obk** [comment]
Seems like all the crates affected are abandoned (which would explain why they haven't been fixed).

@bors r+

**Pull request merged**